google-cloud-pubsub/src/test/java/com/google/cloud/pubsub/v1/OpenTelemetryTest.java

```diff
@@ -0,0 +1,745 @@
+
+    Tracer tracer = openTelemetryTesting.getOpenTelemetry().getTracer("test");
+
+    messageWrapper.startSubscriberSpan(tracer, EXACTLY_ONCE_ENABLED);
+    Span subscribeModackRpcSpan =
+        OpenTelemetryUtil.startSubscribeRpcSpan(
+            tracer,
+            FULL_SUBSCRIPTION_NAME.toString(),
+            "modack",
+            messageWrappers,
+            ACK_DEADLINE,
+            true,
+            true);
+    Span subscribeAckRpcSpan =
+        OpenTelemetryUtil.startSubscribeRpcSpan(
+            tracer, FULL_SUBSCRIPTION_NAME.toString(), "ack", messageWrappers, 0, false, true);
+    Span subscribeNackRpcSpan =
+        OpenTelemetryUtil.startSubscribeRpcSpan(
+            tracer, FULL_SUBSCRIPTION_NAME.toString(), "nack", messageWrappers, 0, false, true);
+
+    Exception e = new Exception("test-exception");
+    OpenTelemetryUtil.setSubscribeRpcSpanException(
+        subscribeModackRpcSpan, true, ACK_DEADLINE, e, true);
+    OpenTelemetryUtil.setSubscribeRpcSpanException(subscribeAckRpcSpan, false, 0, e, true);
+    OpenTelemetryUtil.setSubscribeRpcSpanException(subscribeNackRpcSpan, true, 0, e, true);
+    messageWrapper.endSubscriberSpan();
+
+    List<SpanData> allSpans = openTelemetryTesting.getSpans();
+    assertEquals(4, allSpans.size());
+    SpanData modackSpanData = allSpans.get(0);
+    SpanData ackSpanData = allSpans.get(1);
+    SpanData nackSpanData = allSpans.get(2);
+    SpanData subscriberSpanData = allSpans.get(3);
+
+    StatusData expectedModackStatus =
+        StatusData.create(StatusCode.ERROR, "Exception thrown on modack RPC.");
+    SpanDataAssert modackSpanDataAssert = OpenTelemetryAssertions.assertThat(modackSpanData);
+    modackSpanDataAssert
+        .hasName(SUBSCRIBE_MODACK_RPC_SPAN_NAME)
+        .hasKind(SpanKind.CLIENT)
+        .hasNoParent()
+        .hasStatus(expectedModackStatus)
+        .hasException(e)
+        .hasEnded();
+
+    StatusData expectedAckStatus =
+        StatusData.create(StatusCode.ERROR, "Exception thrown on ack RPC.");
+    SpanDataAssert ackSpanDataAssert = OpenTelemetryAssertions.assertThat(ackSpanData);
+    ackSpanDataAssert
+        .hasName(SUBSCRIBE_ACK_RPC_SPAN_NAME)
+        .hasKind(SpanKind.CLIENT)
+        .hasNoParent()
+        .hasStatus(expectedAckStatus)
+        .hasException(e)
+        .hasEnded();
+
+    StatusData expectedNackStatus =
+        StatusData.create(StatusCode.ERROR, "Exception thrown on nack RPC.");
+    SpanDataAssert nackSpanDataAssert = OpenTelemetryAssertions.assertThat(nackSpanData);
+    nackSpanDataAssert
+        .hasName(SUBSCRIBE_NACK_RPC_SPAN_NAME)
+        .hasKind(SpanKind.CLIENT)
+        .hasNoParent()
+        .hasStatus(expectedNackStatus)
+        .hasException(e)
+        .hasEnded();
+  }
+
+  private PubsubMessage getPubsubMessage() {
+    return PubsubMessage.newBuilder()
+        .setData(ByteString.copyFromUtf8("test-data"))
+        .setOrderingKey(ORDERING_KEY)
+        .build();
+  }
+}
```

Comment:
Do you have any way of testing the performance degradation of this change when EnableOpenTelemetry=false? We want to make sure that users not interested in tracing are not impacted. I suspect the changes are fairly small since we aren't creating any spans, but it would be nice to see the impact of the extra data structures / wrappers.

Reply:
I've done some initial testing to see the impact of adding OpenTelemetry on performance, but will update with results after some more extensive testing.

Reply:
Finished with the refactor/architectural changes. Will run testing again in the coming days and upload results and samples.

Reply:
Attaching tests here as a text file. The difference between the current version the otel version with otel disabled is negligible. We expectedly see increases for the publisher and subscriber when otel is enabled, though more significantly for the subscriber.

[otel_tests.txt](https://github.com/user-attachments/files/17055193/otel_tests.txt)